src/librustc_mir/transform/uninhabited_enum_branching.rs

```diff
@@ -99,26 +99,21 @@ impl<'tcx> MirPass<'tcx> for UninhabitedEnumBranching {
             if let TerminatorKind::SwitchInt { values, targets, .. } =
                 &mut body.basic_blocks_mut()[bb].terminator_mut().kind
             {
-                let vals = &*values;
-                let zipped = vals.iter().zip(targets.iter());
-
-                let mut matched_values = Vec::with_capacity(allowed_variants.len());
-                let mut matched_targets = Vec::with_capacity(allowed_variants.len() + 1);
-
-                for (val, target) in zipped {
-                    if allowed_variants.contains(val) {
-                        matched_values.push(*val);
-                        matched_targets.push(*target);
-                    } else {
-                        trace!("eliminating {:?} -> {:?}", val, target);
-                    }
-                }
-
-                // handle the "otherwise" branch
-                matched_targets.push(targets.pop().unwrap());
-
-                *values = matched_values.into();
-                *targets = matched_targets;
+                // take otherwise out early
+                let otherwise = targets.pop().unwrap();
+                assert_eq!(targets.len(), values.len());
+                let mut i = 0;
+                targets.drain_filter(|_| {
+                    let should_rm = !allowed_variants.contains(&values[i]);
+                    i += 1;
+                    should_rm
+                });
+                targets.push(otherwise);
+
+                values
+                    .to_mut()
+                    .drain_filter(|var| !allowed_variants.contains(var))
+                    .for_each(|_| {});
```

Comment:
uh... you also need to consume the `drain_filter` iterator invoked on `targets`, otherwise it doesn't do anything.

Maybe just use `retain` in both cases?

Reply:
I think when it's dropped, it will still perform the drain filter, but I guess it's just annoying to write it that way